src/main/js/dynamic/DistributionConfiguration.js

```diff
@@ -0,0 +1,315 @@
+            const providerKeys = FieldMapping.retrieveKeys(newProvider.fields);
+            const emptyProviderConfig = FieldModelUtilities.createEmptyFieldModel(providerKeys, newProvider.context, newProvider.name);
+            this.setState({
+                providerConfig: emptyProviderConfig,
+                currentProvider: newProvider
+            });
+        }
+    }
+
+    // FIXME remove this in 5.0.0 as we can change the filterByProject flag to be more appropriate
+    flipFilterFlag(config) {
+        return !FieldModelUtilities.getFieldModelBooleanValueOrDefault(config, KEY_FILTER_BY_PROJECT, false);
+    }
+
+    handleClose() {
+        this.setState({ show: false });
+        this.props.onModalClose();
+        this.props.handleCancel();
+    }
+
+    handleChannelChange({ target }) {
+        const { type, name, value } = target;
+        const updatedValue = type === 'checkbox' ? target.checked.toString() : value;
+        const newState = Array.isArray(updatedValue) ? FieldModelUtilities.updateFieldModelValues(this.state.channelConfig, name, updatedValue) : FieldModelUtilities.updateFieldModelSingleValue(this.state.channelConfig, name, updatedValue);
+
+        this.setState({
+            channelConfig: newState
+        });
+    }
+
+    handleProviderChange({ target }) {
+        const { type, name, value } = target;
+        const updatedValue = type === 'checkbox' ? target.checked.toString() : value;
+        const newState = Array.isArray(updatedValue) ? FieldModelUtilities.updateFieldModelValues(this.state.providerConfig, name, updatedValue) : FieldModelUtilities.updateFieldModelSingleValue(this.state.providerConfig, name, updatedValue);
+
+        this.setState({
+            providerConfig: newState
+        });
+    }
+
+    buildJsonBody() {
+        const { channelConfig, providerConfig } = this.state;
+        const includeAllProjects = this.flipFilterFlag(providerConfig);
+        const updatedProviderConfig = Object.assign({}, FieldModelUtilities.updateFieldModelSingleValue(providerConfig, KEY_FILTER_BY_PROJECT, includeAllProjects));
+        return Object.assign({}, {
+            fieldModels: [
+                channelConfig,
+                updatedProviderConfig
+            ]
+        });
+    }
+
+    handleTestSubmit(event) {
+        event.preventDefault();
+
+        const jsonBody = this.buildJsonBody();
+        this.props.testDistributionJob(jsonBody);
+    }
+
+    handleSubmit(event) {
+        event.preventDefault();
+        const { jobId } = this.props;
+
+        const jsonBody = this.buildJsonBody();
+        if (jobId) {
+            Object.assign(jsonBody, { jobId });
+            this.props.updateDistributionJob(jsonBody);
+        } else {
+            this.props.saveDistributionJob(jsonBody);
+        }
+    }
+
+    // FIXME this function is probably not necessary
```

Comment:
FYI

Reply:
I will probably clean up this and other redundant code in a later PR. There is already a lot going on here.